## Bienvenidos a prerequisitos con Python - Parte 2
Ahora nos adentraremos poco a poco con python desde lo basico hasta tener un entendimiento aceptable de como hacer flujos de programación.

Anteriormente en la parte 1 vimos operadores aritméticos que son parte básica para manipulación de datos.  Ahora veremos unos ejercicios en clase que nos ayudarán a tener entendimiento de python básico.

In [4]:
# Escriba una expresión que calcule el promedio de...
# 30, 16, 40
n = (40 + 16 + 30) /3
print('Resultado esperado: ', n)
print()

Resultado esperado:  28.666666666666668



Resultado esperado: 28.666666666666668

Se nos presenta el siguiente problema.  
Tenemos una pared de 6 metros de largo y 2.8 metros de alto 
Queremos hacer que esta pared sea solamente de vidrio, en donde cada vidrio mide 0.5 metros cuadrados. Calcular cuantos paquetes de vidrios necesitamos comprar si en cada caja de conjunto de vidrio viene una docena de ellos.

In [9]:
a = 0.5
e = 12
na = (e*a)/2
an = int(na)
print('Resultado esperado: ', na)
print()

Resultado esperado:  3.0



Resultado esperado: 3

¿Qué linea de python está bien formulada?
- print(((1+ 24)) + -10//2)
- print((22 - 12)%(3 + 2))
- print ((2 + 16 + 3) / 17)
- print(8/3 - 6*6)



Bajo el estándar PEP8, serían las líneas 2 y 4, averigue porqué

#### Variables y Operadores de Asignacion

Para designar una variable simplemente le damos un nombre descriptivo, esta está siempre a la derecha y para la asignación se da a la izquierda.

    num = 2
    
De otra manera también podemos tener algo como asignar una variable a otra variable.

    a = 1.5
    b = a
    print(b)

Ahora es su turno para demostrar lo que hemos visto hasta el momento.  Veremos también una sintaxis un poco diferente donde tendremos un numero exponenciado 1e3, que es lo mismo que 1000.

In [24]:
# un recipiente cilindrico de agua tiene (en pie cubicos)
recipiente1 = 1e3
# Se tiene otro recipiente de agua (in cubic metres)
recipiente2 = 1.8e3


# haga las siguientes operaciones
# decremente la variable del recipiente 2 al 20%
m = recipiente2 * 0.20
recipiente2 = recipiente2 - m
# añada la variable de recipiente 1 a recipiente 2
recipiente2 = recipiente2 + recipiente1
# incremente el volumen del recipiente 1 al 5%
m1 = recipiente1 * 0.05
recipiente1 = recipiente1 + m1
# decremente el recipiente 1 por 15%
m2 = recipiente1 * 0.15
recipiente1 = recipiente1 - m2
# decremente el 0.3% del recipiente 1,
m3 = recipiente1 * 0.003
recipiente1 = recipiente1 - m3
# decremente el 1% del recipiente2
m4 = recipiente2 * 0.01
recipiente2 = recipiente2 - m4
# sume ambas variables en el recipiente1
recipiente1 = recipiente1 + recipiente2
# imprima el nuevo valor de las variables
print('Resultado esperado: ', recipiente1, recipiente2)

Resultado esperado:  3305.4224999999997 2415.6


Resultado esperado:  3281.325 2440.0

#### Enteros y Flotantes

Refrescando tenemos que una expresión evaluada como 3/4 da un resultado como 0.75.  Los datos más comunes son int y float, la diferencia es que uno es entero y el último incluye decimales.

Se pueden sumar un entero y un flotante, en este caso el resultado será un número flotante.

Adicionalmente podemos convertir de enteros a flotantes y de flotantes a enteros

In [ ]:
3+2.5

In [ ]:
int(3+2.5)

In [ ]:
2 + float(2)

In [ ]:
2+2

In [ ]:
# así tenemos como conseguir el tipo de dato
type(2+2.0)

#### Comparadores booleanos y operadores boleanos

Hemos visto dos tipos de datos, int y float que nos permite realizar operaciones.  En este caso veremos operadores booleanos que solamente sus valores rondan de *Cierto* y *Falso*.

Podemos asignar a una variable un valor booleano como sigue:

    isTrue = True
    isFalse = False

In [ ]:
tengoHambre = True
tengoHambre

In [ ]:
tengoCOVID = False
tengoCOVID

Tambíen podemos evaluar y hacer comparaciones entre operaciones que nos den un resultado booleano:

    <  (menor que)
    >  (mayor que)
    <= (menor o igual que)
    >= (mayor o igual que)
    == (igual que)
    != (no es igual que)

In [ ]:
10 > 2

In [ ]:
2.2 > 2.2

In [ ]:
a = 0.3
b = 0.1 + 0.1 + 0.1
a == b

In [ ]:
a != b

También podemos tener operaciones booleanas que hagan uso de la logica booleana

- and (evalua si se cumplen todas las condiciones)
- or (evalua si alguno es cierto)
- not (invierte un estado booleano)

In [ ]:
tengoSed = True
yaComi = False
esVerdad = False

tengoSed and yaComi and esVerdad

In [ ]:
tengoSed or yaComi or esVerdad

In [ ]:
tengoSed and not yaComi and not esVerdad

Ahora trate de hacer lo siguiente

Escroba un codigo para comparar temperaturas por area

In [11]:
# temperatura de pais vs área de estudio
tempPTY, areaPTY = 38.5, 102
tempCOL, areaCOL = 29.2, 150

# temp/m2
tempm2_pty = tempPTY/areaPTY
tempm2_col = tempCOL/areaCOL

# escriba el restante de código que indica si la temp/m2 de pty es mayor a la temp/m2 de col

print('Resultado esperado: ',tempm2_pty > tempm2_col )

Resultado esperado:  True


Resultado Esperado:  True

##### Pregunta

¿Porqué en Python para comparar no podemos utilizar = sino ==?

#### Strings
Para trabajar con texto en python usamos este tipo de datos.  Podemos crear strings con '' o "" ambos dan una cadena de caracteres.

In [ ]:
'a', "hola"

Los string pueden incluir letras, numeros o caracteres especiales. Pero debemos tener cuidado cuando imprimamos algunos caracteres, por ejemplo el que define una cadena (o string).

In [ ]:
"Nos veremos "tentativamente" a las 1:00p.m. hoy miércoles"

Hay una manera de solucionar este problema o de formatear el texto, podemos hacer esto.

In [ ]:
'Nos veremos "tentativamente" a las 1:00p.m. hoy miércoles'

Pero que sucede si necesitamos ambos tipos de caracteres dentro de evaluar esta expresión.  Es en este caso que usamos '\'.

In [ ]:
'"Nos veremos \"tentativamente\" a las 1:00p.m. hoy miércoles"'

Una vez definida la cadena podemos utilizar algunos operadores para cadenas.
- (+) combinar cadenas
- (\*) repetir cadenas

In [ ]:
inicio = 'Inteligencia Artificial'
final = 'para Internet de Cosas'
inicio + final

Note que une las cadenas pero no las separa por espacio, debemos introducir el espacio nosotros en la cadena o un caracter adicional. Por ejemplo, la respuesta podría ser:
- inicio + ' ' + final

In [ ]:
inicio + ' ' + final

Si queremos repetir el string, podemos utilizar * para esta operación

In [ ]:
inicio*10

Algo importante que queremos hacer a veces es medir el largo de la cadena, esta función embebida nos ayuda a calcular este número.

In [ ]:
data = "0123456789"
len(data)

#### Quiz

Desarrolle lo que se solicita

In [16]:
# TODO:  Arregle lo que está mal

quote = 'Ser-O-No-Ser-He allí-El Dilema'
print(quote)

# LA RESPUESTA ESPERADA DEBE SER:  Ser-O-No-Ser-He allí-El Dilema

Ser-O-No-Ser-He allí-El Dilema


In [ ]:
fruta = "25"
vegetales = "12"

In [21]:
fruta = 25
vegetales =12
total_fruta_y_vegetal = fruta + vegetales
print('Resultado esperado ', total_fruta_y_vegetal, ' y tipo ', type(total_fruta_y_vegetal))

Resultado esperado  37  y tipo  <class 'int'>


Resultado esperado: 37  y tipo  <class 'int'>

Ahora realizaremos un logger sencillo

In [28]:
username = 'Ignacio'
ts = "04:50"
url = "http://minegocio.com/id/nmasaya/me"
print(username,'exited the url ',url ,'at' ,ts)

# TODO: imprimir un mensaje como que se muestra a continuación


Ignacio exited the url  http://minegocio.com/id/nmasaya/me at 04:50


Resulado esperado:

"Ignacio exited the url http://minegocio.com/id/nmasaya/me at 04:50."

In [25]:
nombre = "Juan"
seg_nombre = "Javier"
apellido = "Jimenez"

# calcular el largo del nombre
name_length = len(nombre)

# Ahora que calculó el largo del nombre intente  ver si cabe en la licencia
driving_license_character_limit = 28
print(name_length <= driving_license_character_limit)

True


Resultado esperado:  
True

### Tipos y Tipos de Conversión

Hasta ahora tenemos varios tipos como int, bool, float, string.

El tipo de datos los podemos adquirir con las función type.  Tenemos que utilizar el tipo de datos correspondiente como ya vimos en ejemplos anteriores, de lo contrario no tendremos el tipo de datos correcto para procesar en la aplicación.  

In [ ]:
type(True)

Adicionalmente tendremos casos en donde necesitemos cambiar el tipo de datos, para esto podemos hacer *type casting* y conseguir el dato de salida que necesitemos

In [ ]:
ctr = int(4.0)
print(ctr)
print(type(ctr))

Por otro lado, tendremos situaciones donde deseemos imprimir un entero con una cadena,

In [ ]:
casa = 13
direccion = 'Ave Tesseract-Midgard'
provincia = 'Zanakard'
print(type(casa))
print(type(direccion))

full_dir = str(casa) + ' ' + direccion + ', ' + provincia
print(full_dir)

#### Encuentre los tipos de datos

- "12"
- 12.3
- len('i am a string')

#### Métodos y Cadenas

Hasta ahora hemos visto operadores y funciones.  

Los operadores hasta ahora (más comunes que hemos visto son, entre algunos) +, - , *.

Para el caso de funciones hemos visto algunas como print(), len(), float().

Funciones y operadores tienen el mismo objetivo (modificar o dar resultados de variables), lo único que cambia es como lucen.  Las funciones las podemos identificar rápidamente debido a que poseen paréntesis.

Existen otro tipo de usos con variables llamados métodos.  Los métodos solamente se aplican con un grupo específico de tipo de datos.

In [ ]:
'inteligencia artificial'.title()

Observamos que este método ocupa solo la primera letra de cada frase y la convierte en mayúscula.  Ahora exploremos el siguiente método... 

In [ ]:
materia = 'INTELIGENCIA ARTIFICIAL'
print(materia.islower())

Del ejemplo anterior tenemos que el resultado es falso debido a que no tenemos minúsculas todas, aunque alguna estuviese en mayusculas y las demás en minusculas el resultado sería el mismo.

Los métodos generalmente tienen una palabra que los acompaña entre el paréntesis, estos se llaman argumentos.  Para los métodos utilizados no tienen ningun argumento, pero en realidad está sobreentendido y se lleva como argumento, en este caso, la cadena.

Veamos un ejemplo donde haya un método de más de un argumento.

In [ ]:
print('Abaco 3 Casas 4 Peces 3 Cadenas tres veces treinta y tres veces'.count('tres'))

Aquí vemos cuantas veces se vio aparece la cadena 'tres' existente en esta cadena.

[Vea este vínculo para tener más referencias de cadenas y métodos.](https://www.w3schools.com/python/python_ref_string.asp)

#### Prácticas

- ¿Qué pasa cuando se llama al método islower() para cuando es un punto flotante?
  - Retorna True
  - Retorna False
  - Retorna error

In [13]:
False

False

Del vínculo de métodos intente utilizar alguno

In [14]:
naza = "Hola NaZa , Me Convierte en Minuscula"

x = naza.casefold()

print(x)

hola naza , me convierte en minuscula


In [41]:
# Escriba un código donde asigne un valor a una variable
n ='baile'
a = 'musica'

texto = 'mi pasatiempo favorito es {} , y {}.'.format(n , a)
print(texto)

# Ahora imprima en pantalla con print() pero usando .format() para las variables que se asigno anteriormente e imprima estas dos variables


mi pasatiempo favorito es baile , y musica.


#### Listas

Existen otro tipos de contenedores de datos que son muy útiles para realizar operaciones, las listas.  Las listas son un tipo de datos que se pueden alterar su valor y posición, podemos tanto ordenarlas como establecer su secuencia.

En el siguiente ejemplo veremos un grupo de lista de cadenas.

In [ ]:
materias = ["Robotica", "IA", "Control", "Tecnologia Mecanica"]

Podemos acceder a un elemento de la lista por su índice.

In [ ]:
materias[0], materias[2], materias[-1]

¿Notó algo raro?  Podemos especificar un índice negativo!.  Python permite esto y el índice negativo indica que se empieza desde el final.

¿Qué sudecería si especifico un índice mayor al que posee la lista?.  Simplemente da error porque este índice no está especificado en dicha lista.

In [ ]:
materias[5]

La mayoría de los lenguajes de programación indexan desde cero, siendo cero el primer elemento.

Podemos acceder una parte de un arreglo o lista con python.  Considere el siguiente ejemplo

In [ ]:
valores = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']

primeros_tres = valores[:3]
print(primeros_tres)
primeros_tres = valores[0:3]
print(primeros_tres)

Como observamos anteriormente podemos obviar el primer indice si empezamos desde el primero

In [ ]:
ultimos_tres = valores[5:]
print(ultimos_tres)
ultimos_tres = valores[5:8]
print(ultimos_tres)

También podemos especificar una parte del arreglo

In [ ]:
intermedio = valores[2:6]
intermedio

Existe diferencia entre el largo de la cadena y el largo del arreglo, lo veremos con un ejemplo

In [ ]:
calificacion = 'Su calificación aprueba el curso'
items = ['Pasta', 'Cepillo', 'Bolsa de Basura']

print(len(calificacion))
print(len(items))

Note también que podemos hacer 'slicing' de strings.

In [ ]:
calificacion[1:18]

Un operador muy útil que hará referencia siempre son los operadores de miembros 'in' y 'not in'.
- in : evalua si un objeto existe dentro de la lista
- not in : evalua si un objeto no existe dentro de la lista

In [ ]:
mi_lista = 'Estaremos a la una de la tarde para las clases'
'ar' in mi_lista

In [ ]:
nombres = ['Maria', 'Ana', 'Estela', 'juan']
print('Juan' in nombres, 'Juan' not in nombres)

¿Cómo son diferentes las cadenas de las listas?.  Las listas son un grupo de caracteres, pero las listas son un grupo de ellas, además que podemos cambiarlas.

Las cadenas son un tipo de dato inmutable, es decir, no se puede cambiar.  Las listas son mutables, es decir podemos realizar cambios sobre ellas

In [ ]:
m = 'abc'
m[0] = 'c'

In [ ]:
n = ['xwb', 'abc', 'def']
print(n)
n[0] = 'wxy'
print(n)

#### Ahora intentelo ud.

In [44]:
mes = 'agosto'
dias_del_mes = [31,28,31,30,31,30,31,31,30,31,30,31]

# Use indexing para determinar el número de días en el mes 8
num_dias =  dias_del_mes[7]
print(num_dias)

31


Repuesta esperada: 31

In [46]:
dias = ['Feb 8, 1980', 'Dic 8, 1949', 
        'Mar 20, 1492', 'Abril 12, 1999',
        'Jun 12, 2010']
                 
# Imprima los ultimos tres elementos de la lista
ultimos = dias [2:5]
print(ultimos)

['Mar 20, 1492', 'Abril 12, 1999', 'Jun 12, 2010']


Respuesta esperada: ['Mar 20, 1492', 'Abril 12, 1999', 'Jun 12, 2010']

Suponga que tenemos las siguientes expresiones:

p1 = "Quisiera que se terminara la materia"
p2 = ["Quisiera", "que", "se", "terminara", "la", "materia"]

¿Qué sucede cuando evaluamos estos casos con cada una de las palabras p1 y p2?
- p2[4] = "que"
- p2[0]= "Quisira que"
- p1[30]="a"
- p2[0:2] = ["Que", "quisiera"]

#### Más de listas e inmutabilidad

Dijimos que las listas son mutables y que las cadenas no lo son, veamos un poco más de esto con un ejemplo.

In [ ]:
nombre = 'Coimbra'
jugador = nombre
nombre = 'Pepe'
print(nombre)
print(jugador)

Ahora veamos el mismo caso pero con las listas

In [ ]:
lista_1 = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']
lista_2 = lista_1
lista_1[3] = 'j'
print(lista_1)
print(lista_2)

Notamos entonces que esto hace efecto en ambos casos para ambas variables. 

Existen también métodos para las listas, exploremos max() que nos da el máximo de la lista de objetos dependiendo de cuál tipo de objeto hallamos especificado.

In [ ]:
lista_1 = [2, 14, 20, -1, -5, 4]

print(max(lista_1))

In [ ]:
lista_2 = ['Aries Mu', 'Cygnus Hyoga', 'Pegasus Seiya', 'Pegasus Tenma', 'Sagitarius Aiolos', 'Libra Dohko']
print(max(lista_2))

En las listas de cadenas, la función máx nos dá el resutaldo de la palabra con la que inicia y es mayor dentro del alfabeto.  Es decir, el máximo es el que aparece como máximo alfabeticamente.

In [ ]:
max([12, 'Shaka'])

Aquí nos da error, estamos comparando un entero y un string, para hacer comparaciones se necesita que sean el mismo tipo de datos.

Veamos ahora el método sorted(), que nos da una lista ordenada dependiendo del tipo de datos.

In [ ]:
numbers = [24, 48, 54, 65, 24, 12, 1]
print(sorted(numbers))

In [ ]:
days_of_week = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
print(sorted(days_of_week, reverse=True))

Otro método muy útil es join().  Este nos da la posibilidad de unir, en este caso una cadena por medio de caracteres específicos u otro string adicional.

In [ ]:
"-".join(['No', 'te', 'voy', 'a', 'decir'])

Es importante que cuando estemos utilizando join(), el arreglo sea separado por comas, si no se hace esto no dará error pero si un resultado no deseado y difícil de depurar.

In [ ]:
"-".join(['No' 'te' 'voy' 'a' 'decir'])

Si intentamos unificar tanto una cadena como un número esto nos dará error.

In [ ]:
" ".join(['No', 42, 'es' 'un' 'número'])

Otro método util para arreglos es append().  Veamos un ejemplo.

In [105]:
a = ['1', '2', '3']
a.append('5')
print(a)

['1', '2', '3', '5']


#### Intentelo ahora ud.

¿Cuál será la salida del siguiente código?

In [47]:
a = [1, 5, 8]
b = [2, 6, 9, 10]
c = [100, 200]

print(max([len(a), len(b), len(c)]) , end=" ")
print(min([len(a), len(b), len(c)]))

4 2


Respuesta esperada:
4
2

¿Cuál será la salida del siguiente código?

In [ ]:
names = ["Uno", "Dos", "Tres", "Cuatro"]
print(" & ".join(sorted(names)))

Respuesta esperada:

Cuatro & Dos & Tres & Uno

¿Cuál será la salida del siguiente código?

In [ ]:
names = ["Norberto", "Ignacio", "Luis", "Tito"]
names.append("Armando")
print(sorted(names))

Respuesta esperada:

['Armando', 'Ignacio', 'Luis', 'Norberto', 'Tito']

#### Tuples

Python provee otro contenedor de datos, los tuples.  Es una estructura de datos que es inmutable.  Concuerdan con las listas que los datos se guardan ordenadamente y se pueden acceder por medio de índices.

In [98]:
lat_long = (24.21254, 145.04011)

print(lat_long)
print('La latitud  es: {}'.format(lat_long[0]))
print('La longitud es: {}'.format(lat_long[1]))

(24.21254, 145.04011)
La latitud  es: 24.21254
La longitud es: 145.04011


Como dijimos los tuples son inmutables, no podemos modificarlos, hacer operaciones de suma entre tuples.  Normalmente los tuples se utilizan como datos que son correlacionados entre uno y el otro y sabemos que siempre será así, como por ejemplo la latitud y longitud o también sexo y edad por ejemplo, que siempre están relacionados a la persona.

Podemos también asignar los tuples de manera compacta, vea el siguiente ejemplo.

In [100]:
tesseract = 12, 14, 20, 34
D1, D2, D3, D4 = tesseract
print('Las dimensiones del tesseract son {}x{}x{}x{}'.format(D1,D2,D3,D4))
print(f'Las dimensiones del tesseract son {D1}x{D2}x{D3}x{D4}')

Las dimensiones del tesseract son 12x14x20x34
Las dimensiones del tesseract son 12x14x20x34


#### Intentelo ud.

¿Cuál será la salida del siguiente código?

In [ ]:
tuple_a = 2, 8
tuple_b = (8, 2)

print(tuple_a == tuple_b)
print(tuple_a[1])

Resultado esperado:

False
8

#### Sets

Imaginemos que tenemos una aplicacion donde se conectan los usuarios y estos usuarios son de diferentes partes del mundo, tendremos entonces por ejemplo, una lista de miles de usuarios de diferentes regiones del planeta y los paises se verán duplicados, es aquí donde un 'set' es útil.

Si recuerda, las listas generaban una salida incluso cuando el mismo dato era repetido, este aún existia.

Podemos crear un 'set' de una 'lista' o 'array' si hacemos lo siguiente

In [104]:
nombres = ['Ana', 'Gabriel', 'Juan', 'Laura', 'Roberto', 'Laura', 'Nestor', 'Ana', 'Cecibel']
print(nombres, len(nombres))
print(set(nombres), len(set(nombres)))

['Ana', 'Gabriel', 'Juan', 'Laura', 'Roberto', 'Laura', 'Nestor', 'Ana', 'Cecibel'] 9
{'Gabriel', 'Cecibel', 'Ana', 'Roberto', 'Nestor', 'Juan', 'Laura'} 7


También tendremos casos en donde querramos añadir datos al final del set o remover del set un dato.  Centrémonos solamente en el ultimo dato insertado o la extracción del primer dato del set.

In [109]:
my_set = set([1, 2, 4, 5, 19, 20])
print(my_set)
my_set.add(-1)
print(my_set)
my_set.pop()
print(my_set)

{1, 2, 4, 5, 19, 20}
{1, 2, 4, 5, 19, 20, -1}
{2, 4, 5, 19, 20, -1}


#### Ahora intentelo ud.

¿Cuál será la salida del siguiente código?

In [48]:
a = [1, 2, 2, 4, 4, 4, 8, 8, 8, 8]
b = set(a)
print(len(a) - len(b)) 

6


Respuesta esperada:

6

Considere el siguiente código.

In [111]:
a = [0, -1, -1, 2, 1, 2, 7, 9, 12, 15]
b = set(a)
b.add(111)
b.pop()

1

Después de ejecutar el código, ¿el número 111 seguirá siendo parte del set?

#### Diccionarios y operadores de identidades

Sets son estructuras simples de datos con la opción de coleccinar un solo tipo de datos.  Los diccionarios operan como los sets pero podemos coleccionar diferentes tipos de datos por 'keys' y 'values'.

Para un diccionario 'key' es el equivalente de el índice en la lista, solo que este puede ser cualquier tipo de dato. 

'values' son los valores de esa específica llave ('key'), que es por el que se accesará el elemento.

In [114]:
my_dict = {'January':1, 'February':2, 'March':3}

print(my_dict['February'])
print(my_dict)

2
{'January': 1, 'February': 2, 'March': 3}


No necesariamente podemos tener el 'key' del mismo tipo, pero normalmente acostumbro a uniformar este tipo de 'key'.

También podemos utilizar el referente 'in' para buscar si un elemento se encuentra dentro del diccionario.

In [115]:
my_dict = {'January':1, 'February':2, 'March':3, 'April':4, 'May':5, 'June':6, 'July':7, 
           'August':8, 'September':9, 'October':10, 'November':11, 'December':12}
print(my_dict)
print('Monday' in my_dict)

{'January': 1, 'February': 2, 'March': 3, 'April': 4, 'May': 5, 'June': 6, 'July': 7, 'August': 8, 'September': 9, 'October': 10, 'November': 11, 'December': 12}
False


Una manera de extraer valores del diccionario es utilizando get().  Veamos un ejemplo.

In [117]:
print(my_dict.get('September'))
print(my_dict.get('Monday'))
print(my_dict['Monday'])

9
None


KeyError: 'Monday'

Observó que para el primer caso extrajo el valor dependiendo de la llave en el diccionario y para el caso siguiente, como no estaba dentro de la posiblidad entonces nos dio *None* como respuesta.

Para el último caso notamos que buscamos una llave que no existe utilizando el nombre específico de esta llave y el programa nos da error, esto puede causar problemas en su código pero no es una mala práctica, adopte la mejor para ud.

Podemos corregir este inconveniente verificando, como en el siguiente código.

In [119]:
mes = my_dict.get('June')
not_null = mes is not None
print(not_null)

True


Introducimos en el código anterior operadores de identidad...
- is
- is not

Veremos más en el siguiente quiz.

### Intentelo ud ahora

Defina un diccionario de los siguientes datos

Keys, Values
- Panama, 5.9
- Costa Rica, 6.2
- Colombia, 8.4
- Mexico, 1.2

¿Cuáles de las siguientes pueden ser utilizadas como diccionario?
- str
- list
- int
- float

In [53]:
str, float, int

(str, float, int)

¿Qué sucede si buscamos un valor que no existe en un diccionario?

Para verificar, realizar:
- Crear un diccionario
- Buscar por braquets un valor no existente
- Qué sucede

In [57]:
mi_dicc = {'Panama':5.9, 'Costa Rica': 6.2, 'Colombia': 8.4 'Mexico': 1.2}
print(mi_dicc)
print(mi_dicc.get('Argentina'))

SyntaxError: invalid syntax (<ipython-input-57-c99cd99917bf>, line 1)

Tenemos el siguiente diccionario...
¿Cómo accedemos al tercer elemento de cada llave?

In [1]:
items = {'harina': [20, 10, 15, 8, 32, 15], 
 'carne': [3,4,2,8,2,4], 
 'pan': [2, 3, 3], 
 'cc': [0.3, 0.5, 0.8, 0.3, 1]}

Respuesta esperada:

(15, 2, 3, 0.8)

#### Estructura de Datos compuesta

Nosotros podemos hacer más con diccionarios, podemos tener una estructura compuesta. ¿Para que sirve esto? Imaginemos que tenemos un grupo de datos de cada persona y queremos almacenar su sexo y su peso, esto no lo podemos realizar con simplemente una llave y un valor.  Es aquí donde realizamos la composición de estos datos.

Podemos entonces tener un diccionario dentro de otro.

In [123]:
datos = {'Ana':{'edad':24, 'peso':110, 'sexo':'F'},
         'Javi':{'edad':12, 'peso':90, 'sexo':'M'}}
print(datos)
print(datos.get('Javi'))
print(datos.get('Juan'), 'No existe Juan')

{'Ana': {'edad': 24, 'peso': 110, 'sexo': 'F'}, 'Javi': {'edad': 12, 'peso': 90, 'sexo': 'M'}}
{'edad': 12, 'peso': 90, 'sexo': 'M'}
None No existe Juan


También podemos buscar el tipo de datos como sigue:

In [124]:
datos['Javi']['peso']

90

#### Intentelo ud ahora

- Añada una nueva entrada a cada subgrupo del diccionario llamada 'estudiante' al diccionario 'datos'
- Trate de buscar por medio de estos valores y conseguir las siguientes respuestas
  - datos.get['Ana']['estudiante']
    - True
  - datos.get['Javi']['estudiante']
    - False

### Fin de la parte de prerequisitos 2